temporal-serviceclient/src/main/java/io/temporal/serviceclient/ChannelManager.java

```diff
@@ -0,0 +1,341 @@
+ *  located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ *  or in the "license" file accompanying this file. This file is distributed on
+ *  an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ *  express or implied. See the License for the specific language governing
+ *  permissions and limitations under the License.
+ */
+
+package io.temporal.serviceclient;
+
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import io.grpc.*;
+import io.grpc.health.v1.HealthCheckRequest;
+import io.grpc.health.v1.HealthCheckResponse;
+import io.grpc.health.v1.HealthGrpc;
+import io.grpc.netty.shaded.io.grpc.netty.NettyChannelBuilder;
+import io.grpc.stub.MetadataUtils;
+import io.temporal.internal.retryer.GrpcRetryer;
+import java.util.Collection;
+import java.util.List;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+class ChannelManager {
+  private static final Logger log = LoggerFactory.getLogger(ChannelManager.class);
+
+  private static final int MAX_INBOUND_MESSAGE_SIZE = 25_000_000;
+
+  /** refers to the name of the gRPC header that contains the client library version */
+  private static final Metadata.Key<String> LIBRARY_VERSION_HEADER_KEY =
+      Metadata.Key.of("client-version", Metadata.ASCII_STRING_MARSHALLER);
+
+  /** refers to the name of the gRPC header that contains supported server versions */
+  private static final Metadata.Key<String> SUPPORTED_SERVER_VERSIONS_HEADER_KEY =
+      Metadata.Key.of("supported-server-versions", Metadata.ASCII_STRING_MARSHALLER);
+
+  /** refers to the name of the gRPC header that contains the client SDK name */
+  private static final Metadata.Key<String> CLIENT_NAME_HEADER_KEY =
+      Metadata.Key.of("client-name", Metadata.ASCII_STRING_MARSHALLER);
+
+  private static final String CLIENT_NAME_HEADER_VALUE = "temporal-java";
+
+  private final ServiceStubsOptions options;
+
+  private final AtomicBoolean shutdownRequested = new AtomicBoolean();
+  // Shutdown channel that was created by us
+  private final boolean channelNeedsShutdown;
+  private final ScheduledExecutorService grpcConnectionManager;
+
+  protected final ManagedChannel channel;
+  protected final Channel interceptedChannel;
+  protected final HealthGrpc.HealthBlockingStub healthBlockingStub;
+
+  public ChannelManager(
+      ServiceStubsOptions options, List<ClientInterceptor> additionalHeadInterceptors) {
+    // Do not shutdown a channel passed to the constructor from outside
+    this.channelNeedsShutdown = options.getChannel() == null;
+
+    this.options = options;
+    if (options.getChannel() != null) {
+      this.channel = options.getChannel();
+      this.grpcConnectionManager = null;
+    } else {
+      this.channel = prepareChannel();
```

Comment:
To confirm, by default we are intentionally creating separate channels for different gRPC services on the same server? I haven't put in thought to how the API would look if you wanted to share them, but I just want to make sure that is intentional.

Reply:
Yes. By default, every stub will have its own channel. It happens now if we create several WorkflowServiceStubs too. At the same time, if the users really want, they can reuse the channel. I exposed ServiceStubs#getRawChannel and the channel can be taken from the existing stubs and reused in the Options for a new stub.

Also, keep in mind that a regular App shouldn't be creating OperatorServiceStubs at all.